README.md
## DDE Control Center

DDE Control Center is the control panel of Deepin Desktop Environment.
## Dependencies

### Build dependencies
* pkg-config
* cmake
* Qt5(>= 5.6) with modules:
  - qt5-default
  - qt5declarative5-dev
  - qt5svg5-dev
  - qttools5-dev-tools
  - qtdeclarative5-dev
* Deepin-tool-kit(>=2.0) with modules:
  - dtkcore-dev
  - dtkwidget-dev
  - dtkcore-bin
* geoip-dev
* dde-qt-dbus-factory
* gsettings-qt-dev
* kf5networkmanagerqt-dev
* xrandr-dev
* mtdev-dev
* fontconfig1-dev
* freetype6-dev
* egl1-mesa-dev
* xcb-ewmh-dev

### Runtime dependencies

* [deepin-tool-kit](https://github.com/linuxdeepin/deepin-tool-kit)
* [dde-api](https://github.com/linuxdeepin/dde-api)
* [dde-daemon](https://github.com/linuxdeepin/dde-daemon)
* [startdde](https://github.com/linuxdeepin/startdde)
* [dde-qt-dbus-factory](https://github.com/linuxdeepin/dde-qt-dbus-factory)
* [qt5integration](https://github.com/linuxdeepin/qt5integration)
* geoip-database
* Qt5(>= 5.6)
  * Qt5-DBus
  * Qt5-Multimedia
  * Qt5-MultimediaWidgets
  * Qt5-Svg

## Installation

### Build from source code

1. Make sure you have installed all dependencies.

2. Build:
```
$ cd dde-control-center
$ mkdir Build
$ cd Build
$ cmake ..
$ make
```

3. Install:
```
$ sudo make install
```

The executable binary file could be found at `/usr/bin/dde-control-center` after the installation is finished, and plugins will be placed into `/usr/lib/dde-control-center/modules/`.

## Usage

Execute `dde-control-center -h` to get more details.

## Getting help

You can press `F1` to start [deepin-manual](https://github.com/linuxdeepin/deepin-manual) when you focus on DDE Control Center window.

You may also find these channels useful if you encounter any other issues:

* [Gitter](https://gitter.im/orgs/linuxdeepin/rooms)
* [IRC Channel](https://webchat.freenode.net/?channels=deepin)
* [Official Forum](https://bbs.deepin.org/)
* [Wiki](https://wiki.deepin.org/)

## Getting involved

We encourage you to report issues and contribute changes

* [Contribution guide for developers](https://github.com/linuxdeepin/developer-center/wiki/Contribution-Guidelines-for-Developers-en). (English)
* [开发者代码贡献指南](https://github.com/linuxdeepin/developer-center/wiki/Contribution-Guidelines-for-Developers) (中文)

## License

DDE Control Center is licensed under [GPLv3](LICENSE).


CHANGELOG.md
<a name="4.10.12"></a>
### 4.10.12 (2019-06-25)


#### Bug Fixes

* **ldap:**  login not enabled after joining domain ([0c645d91](https://github.com/linuxdeepin/dde-control-center/commit/0c645d91055b488d1d6bde0dd7f732bb7e3be2ff))
* **nav:**  missing disable systeminfo module ([02511c4a](https://github.com/linuxdeepin/dde-control-center/commit/02511c4ae7f10887dceb503f4d6a7d6db745d919))
* **sync:**  typo DISALBE -> DISABLE ([95356783](https://github.com/linuxdeepin/dde-control-center/commit/95356783358dfa0581122d9da50eb6b6c2b649d8))

#### Others

*   missing build macro ([b8681869](https://github.com/linuxdeepin/dde-control-center/commit/b8681869731642764603db33f748566ff82cdac0))
* **update:**  remove network check ([b52f5267](https://github.com/linuxdeepin/dde-control-center/commit/b52f526753bdd9822934e13d346d70ceb4f0bfd8))

#### Features

* **navbar:**  avoid tip grab focus ([d4251454](https://github.com/linuxdeepin/dde-control-center/commit/d4251454f512a8f0ab0be9aa444b4c24b8124c65))



<a name="4.10.11"></a>
### 4.10.11 (2019-06-14)


#### Bug Fixes

* **accounts:**  cannot modify user password ([0c68cc3b](https://github.com/linuxdeepin/dde-control-center/commit/0c68cc3bd9e2d4b7f161322094c3da734fc5818f))
* **network:**  Abnormal interface of hostpot ([551d13e9](https://github.com/linuxdeepin/dde-control-center/commit/551d13e904c232fae0e193c846f948208fd44da0))



<a name="4.10.10"></a>
### 4.10.10 (2019-06-13)


#### Others

*   update license ([765fc346](https://github.com/linuxdeepin/dde-control-center/commit/765fc346bf2107bd33af55cb46585878187ae375))



<a name="4.10.9"></a>
### 4.10.9 (2019-06-11)


#### Features

* **Sync:**  add config files ([a9415192](https://github.com/linuxdeepin/dde-control-center/commit/a941519291894c2f3fa43248b34cd8e5305f002f))

#### Bug Fixes

*   incorrect layout of various pages. ([c9360985](https://github.com/linuxdeepin/dde-control-center/commit/c9360985d99ee4502c5dd7e04dd4cd19df7cfe23))
*   set word wrap on switch widget construction. ([bee7372e](https://github.com/linuxdeepin/dde-control-center/commit/bee7372e84deec36c4d83a5f2c1ec1d00805e79f))
* **sync:**
  *  syncstate widget was truncated the first time started ([9c814dd9](https://github.com/linuxdeepin/dde-control-center/commit/9c814dd979feef9d08fe74649fd64b0500466741))
  *  last sync time always not show ([e29491fe](https://github.com/linuxdeepin/dde-control-center/commit/e29491fecb1f5a95dc22fb979ee4035b722aea0c))
* **update:**  missing smarttips spacing ([f9e5ff78](https://github.com/linuxdeepin/dde-control-center/commit/f9e5ff7812610aeafcde2413e2d3d636940a1724))



<a name="4.10.8"></a>
### 4.10.8 (2019-06-04)


#### Bug Fixes

*   switch widget caption text overflows. (#212) ([a16647b3](https://github.com/linuxdeepin/dde-control-center/commit/a16647b30591978721fdc5d7f61345970b7a60e9))
*   dbus method 'rect' returns wrong results after screen rotation. ([6fe8c17d](https://github.com/linuxdeepin/dde-control-center/commit/6fe8c17dc4386dc9f0093a6121c462c0c445ed31))
* **dbus:**  check windowhandle ([25f0ff7d](https://github.com/linuxdeepin/dde-control-center/commit/25f0ff7dd7e931991d3695feaab0551db2a5477d))

#### Features

* **sync:**
  *  add region check ([c519e4b4](https://github.com/linuxdeepin/dde-control-center/commit/c519e4b4e8b5402a22b65e8b23e1feb7fdb558e8))
  *  disable cloud sync for professional platform ([1e6c4dee](https://github.com/linuxdeepin/dde-control-center/commit/1e6c4deebfc0d099ded972ef8b0042706b9981c6))
  *  upload new syncing icon ([f5eed42a](https://github.com/linuxdeepin/dde-control-center/commit/f5eed42a0abda087abafe9f09f94085d0a0a56bb))
  *  check sync daemon isvalid ([9da7475a](https://github.com/linuxdeepin/dde-control-center/commit/9da7475a0352136666d1612e0618fdb33eac597b))
  *  set syncing when lastsynctime is zero ([1f89546d](https://github.com/linuxdeepin/dde-control-center/commit/1f89546de52aac1b369134249d06e2492e2fc764))
  *  add sync module state changed ([ae369562](https://github.com/linuxdeepin/dde-control-center/commit/ae3695624d96223b75751f48a6001fbeee9bb4e7))
  *  add logout button ([550652ca](https://github.com/linuxdeepin/dde-control-center/commit/550652ca260442f26a17194e672249792f8c4ce8))
  *  add sync state icon ([88eba3c8](https://github.com/linuxdeepin/dde-control-center/commit/88eba3c8c58d12858f5c60ea94e7e9cfc923280b))
  *  add theme power corner SyncType ([eff81f96](https://github.com/linuxdeepin/dde-control-center/commit/eff81f96c70324838df2a4f0f3f5dcab449d49b9))
  *  add sync module list ([d02993d5](https://github.com/linuxdeepin/dde-control-center/commit/d02993d566862dc8d28150f6f818029bb09c651a))
  *  add sync state widget ([aa5a5bdf](https://github.com/linuxdeepin/dde-control-center/commit/aa5a5bdf71f1ce237e64019b2a497088c789b5f7))
  *  show user info when user logind ([7d36e283](https://github.com/linuxdeepin/dde-control-center/commit/7d36e283c8fc767efebd44b748fd84adedb1f9b7))
  *  init sync module ([9a8ae08d](https://github.com/linuxdeepin/dde-control-center/commit/9a8ae08d660cf9ec5d5e85af60207ed31aec3218))



<a name="4.10.7"></a>
### 4.10.7 (2019-06-01)


#### Bug Fixes

*   long processor name screws up layout. ([722a95ef](https://github.com/linuxdeepin/dde-control-center/commit/722a95efc9e4d85fe83683ae88927eaaeb4436b6))
* **fprintd:**  refresh device when enter user detail page again ([b912e969](https://github.com/linuxdeepin/dde-control-center/commit/b912e969fea91c3c1f1baf6f5b5ea9ac077304ff))



<a name="4.10.6"></a>
### 4.10.6 (2019-05-29)


#### Bug Fixes

* **accounts:**  not hide error tip ([c4818bc7](https://github.com/linuxdeepin/dde-control-center/commit/c4818bc7dd9c60415f2fbaf0c2a2bf4383e40310))
* **sw:**  missing mieee ([048511fd](https://github.com/linuxdeepin/dde-control-center/commit/048511fdc85714d13cb77a6ee2d12269a2c3c5f3))



<a name="4.10.5"></a>
### 4.10.5 (2019-05-23)


#### Bug Fixes


frame/main.cpp
/*
 * Copyright (C) 2015 ~ 2018 Deepin Technology Co., Ltd.
 *
 * Author:     sbw <sbw@sbw.so>
 *             kirigaya <kirigaya@mkacg.com>
 *             Hualet <mr.asianwang@gmail.com>
 *
 * Maintainer: sbw <sbw@sbw.so>
 *             kirigaya <kirigaya@mkacg.com>
 *             Hualet <mr.asianwang@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "frame.h"
#include "dbuscontrolcenterservice.h"

#include <DApplication>
#include <DDBusSender>
#include <DLog>
#include <QStyle>

DWIDGET_USE_NAMESPACE
DCORE_USE_NAMESPACE

static const QString getQssFromFile(const QString &name)
{
#ifdef QT_DEBUG
//    qDebug() << "load qss file: " << name;
#endif

    QString qss;

    QFile f(name);
    if (f.open(QIODevice::ReadOnly | QIODevice::Text)) {
        qss = f.readAll();
        f.close();
    }

    return qss;
}

static const QString getStyleSheetFromDir(QDir dir)
{
    QString ret;

    for (auto name : dir.entryInfoList(QDir::Files | QDir::NoDotAndDotDot)) {
        if (name.suffix() == "qss") {
            ret.append(getQssFromFile(name.absoluteFilePath()));
        }
    }

    return ret;
}

static const QString styleSheetFromTheme(const QString &theme)
{
    QStringList moduleList = { "frame", "widgets", "accounts", "defapp",
                               "mouse", "wacom", "sound", "bluetooth",
                               "personalization", "datetime", "display",
                               "keyboard", "power", "systeminfo",  "update", "network"};

    QString ret;

    const QString resources = ":/%1/themes/" + theme;
    for (auto module : moduleList)
    {
        QString dir = resources.arg(module);
        ret.append(getStyleSheetFromDir(dir));
    }

    return ret;
}

static void onThemeChange(const QString &theme)
{
    QString qss;
    qss.append(styleSheetFromTheme("common"));
    qss.append(styleSheetFromTheme(theme));

#ifdef QT_DEBUG
//    qDebug() << "set theme:" << theme;
//    qDebug().noquote() << "qss: " << endl << qss;
#endif

    qApp->setStyleSheet(qss);
}

static void onFontSizeChanged(const float pointSizeF) {
    // TODO(hualet): DPI default to 96.
    auto PxToPt = [] (int px) -> float {
        return px * 72.0 / 96;
    };

    QFont font(qApp->font());
    if (pointSizeF <= 8.2) {
        font.setPointSizeF(PxToPt(10));
        qApp->setFont(font, "dcc::widgets::SmallLabel");
        font.setPointSizeF(PxToPt(11));
        qApp->setFont(font, "dcc::widgets::NormalLabel");
        font.setPointSizeF(PxToPt(12));
        qApp->setFont(font, "dcc::widgets::LargeLabel");
    } else if (pointSizeF == 9) {
        font.setPointSizeF(PxToPt(11));
        qApp->setFont(font, "dcc::widgets::SmallLabel");
        font.setPointSizeF(PxToPt(12));
        qApp->setFont(font, "dcc::widgets::NormalLabel");
        font.setPointSizeF(PxToPt(16));
        qApp->setFont(font, "dcc::widgets::LargeLabel");
    }  else if (pointSizeF == 9.7) {
        font.setPointSizeF(PxToPt(11));
        qApp->setFont(font, "dcc::widgets::SmallLabel");
        font.setPointSizeF(PxToPt(13));
        qApp->setFont(font, "dcc::widgets::NormalLabel");
        font.setPointSizeF(PxToPt(15));
        qApp->setFont(font, "dcc::widgets::LargeLabel");
    } else if (pointSizeF == 11.2) {
        font.setPointSizeF(PxToPt(12));
        qApp->setFont(font, "dcc::widgets::SmallLabel");
        font.setPointSizeF(PxToPt(15));
        qApp->setFont(font, "dcc::widgets::NormalLabel");
        font.setPointSizeF(PxToPt(18));
        qApp->setFont(font, "dcc::widgets::LargeLabel");
    } else if (pointSizeF >= 12) {
        font.setPointSizeF(PxToPt(14));
        qApp->setFont(font, "dcc::widgets::SmallLabel");
        font.setPointSizeF(PxToPt(16));
        qApp->setFont(font, "dcc::widgets::NormalLabel");
        font.setPointSizeF(PxToPt(20));
        qApp->setFont(font, "dcc::widgets::LargeLabel");
    }
}

int main(int argc, char *argv[])
{
    DApplication::loadDXcbPlugin();

    DApplication app(argc, argv);
    app.setOrganizationName("deepin");
    app.setApplicationName("dde-control-center");
    app.setApplicationVersion("4.0");
    app.setAttribute(Qt::AA_UseHighDpiPixmaps);
    app.loadTranslator();
//    app.loadTranslator(QList<QLocale>() << QLocale::Chinese);
    app.setQuitOnLastWindowClosed(false);
    app.setTheme("light");

    DLogManager::registerConsoleAppender();
    DLogManager::registerFileAppender();

    // take care of command line options
    QCommandLineOption showOption(QStringList() << "s" << "show", "show control center(hide for default).");
    QCommandLineOption toggleOption(QStringList() << "t" << "toggle", "toggle control center visible.");
    QCommandLineOption moduleOption("m", "the module' id of which to be shown.", "module");
    QCommandLineOption pageOption("p", "specified module page", "page");

    QCommandLineParser parser;
    parser.setApplicationDescription("DDE Control Center");
    parser.addHelpOption();
    parser.addVersionOption();
    parser.addOption(showOption);
    parser.addOption(toggleOption);
    parser.addOption(moduleOption);
    parser.addOption(pageOption);
    parser.process(app);

    const QString &reqModule = parser.value(moduleOption);
    const QString &reqPage = parser.value(pageOption);

    app.setTheme("semidark");
    onThemeChange("dark");

    QTimer::singleShot(0, [] { onFontSizeChanged(qApp->font().pointSizeF()); });

    Frame f;
    DBusControlCenterService adaptor(&f);
    Q_UNUSED(adaptor);
    QDBusConnection conn = QDBusConnection::sessionBus();
    if (!conn.registerService("com.deepin.dde.ControlCenter") ||
        !conn.registerObject("/com/deepin/dde/ControlCenter", &f)) {
        qDebug() << "dbus service already registered!";

        if (parser.isSet(toggleOption))
            DDBusSender()
                    .service("com.deepin.dde.ControlCenter")
                    .interface("com.deepin.dde.ControlCenter")
                    .path("/com/deepin/dde/ControlCenter")
                    .method("Toggle")
                    .call();

        if (!reqModule.isEmpty())
            DDBusSender()
                    .service("com.deepin.dde.ControlCenter")
                    .interface("com.deepin.dde.ControlCenter")
                    .path("/com/deepin/dde/ControlCenter")
                    .method("ShowPage")
                    .arg(reqModule)
                    .arg(reqPage)
                    .call();

        else if (parser.isSet(showOption))
            DDBusSender()
                    .service("com.deepin.dde.ControlCenter")
                    .interface("com.deepin.dde.ControlCenter")
                    .path("/com/deepin/dde/ControlCenter")
                    .method("Show")
                    .call();

#ifndef QT_DEBUG
        return 0;
#endif
    }

#ifndef QT_DEBUG
    if (parser.isSet(showOption) || parser.isSet(toggleOption))
#endif
        QTimer::singleShot(1, &f, &Frame::startup);

    if (!reqModule.isEmpty())
        QTimer::singleShot(10, &f, [&] { f.showSettingsPage(reqModule, reqPage); });

    QObject::connect(&f, &Frame::fontSizeChanged, &f, [=] {
        onFontSizeChanged(qApp->font().pointSizeF());
    });

    return app.exec();
}
